docs/design-documents/20220408-loading-pipeline-configuration-files.md

```diff
@@ -0,0 +1,282 @@
+these concerns, but there are additional design decisions to make.
+
+## Design 
+
+We propose the creation of a pipeline configuration file that contains all of
+the configuration values needed for `n` number of full pipelines.
+
+We propose naming this file `pipelines.yml` and it's default location should be
+at the root of the project. The format of the file is to be determined, though
+`yaml` has been proposed previously.
+
+If this file is passed at start time, we propose that Conduit run in production 
+mode. Production mode initially will not serve the UI. In the future, it would 
+serve a read-only optimized view of Conduit.
+
+Configuration must securely accept secret information. We don't want to solve
+the problem of secrets management, but we must be in the business of secret
+concealment. Additionally, we already allow environment variables so we must 
+continue handling those correctly. 
+
+We propose merging CLI flags, environment variables, and the user's provided 
+configuration file, in that order, at runtime into template variables in the 
+configuration file with their secret values at runtime.
+
+We propose merging environment variables and CLI flags into a set of values 
+that is then injected into the provided set of pipeline files.
+
+We propose supporting Go templating in all pipeline files to allow for dynamic
+injection of those environment variables at runtime.
+
+## Implementation
+
+Conduit currently defines a `Config` struct in `pkg/conduit` that it 
+accepts at runtime to configure behavior. 
+
+We propose modifying this configuration struct to allow for a configuration
+file to be set for Conduit to read.
+
+```go
+// Config holds all configurable values for Conduit.
+type Config struct {
+	// File holds a Path to a configuration file that Conduit should load.
+	File struct {
+		Path string
+	}
+
+	DB struct {
+		Type   string
+		Badger struct {
+			Path string
+		}
+		Postgres struct {
+			ConnectionString string
+			Table            string
+		}
+	}
+
+	HTTP struct {
+		Address string
+	}
+	GRPC struct {
+		Address string
+	}
+
+	Log struct {
+		Level  string
+		Format string
+	}
+}
+```
+
+The above configuration is parsed at start time and provies a reference to our 
+pipeline configuration file. After parsing that file, we inject environment 
+variables into their corresponding templated locations, if any.
+
+```go
+// DB defines the interface for a key-value store.
+type DB interface {
+	// NewTransaction starts and returns a new transaction.
+	NewTransaction(ctx context.Context, update bool) (Transaction, context.Context, error)
+	// Close should flush any cached writes, release all resources and close the
+	// store. After Close is called other methods should return an error.
+	Close() error
+
+	Set(ctx context.Context, key string, value []byte) error
+	Get(ctx context.Context, key string) ([]byte, error)
+	GetKeys(ctx context.Context, prefix string) ([]string, error)
+}
+```
+
+### Configuration driver
```

Comment:
The configuration driver would implement `database.DB` and only allow "read-only" operations, right? I feel like this was written with "production mode" in mind, so this would need to be rethought if we decide we only want to make single pipelines immutable instead of the whole Conduit instance. Even in case of "production mode" the problem from my first comment still persists.